### DISCUSSION TITLE
Unphysical oscillations in tangential forces for contact problem

### DISCUSSION THREAD

**Discussion by npurwar93** [idaholab/moose#24562] [Q&A Modules: Solid mechanics] [answered]
I am trying to understand the contact module of MOOSE. As a starting problem, I have a very simple setup: a cube block (box) pressed and slided over another one-element rigid block (plane) with friction. I make the plane as rigid by fixing all its nodes.

```
[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
  patch_update_strategy = 'iteration'
  volumetric_locking_correction = true
[]

[Mesh]
  [rectangle_body]
    type = GeneratedMeshGenerator
    dim = 3
    xmin= 20
    xmax = 50
    ymin = 6
    ymax = 36
    zmin = -15
    zmax = 15
    nx = 2
    ny = 2
    nz = 2
    elem_type = HEX
    boundary_name_prefix = 'box'
  []

  [plane_rigid]
    type = GeneratedMeshGenerator
    dim = 3
    xmin= 0
    xmax = 100
    ymin = 0
    ymax = 5
    zmin = -50
    zmax = 50
    nx = 1
    ny = 1
    nz = 1
    elem_type = HEX
    boundary_name_prefix = 'plane'
    boundary_id_offset = 6
    subdomain_ids = 1
  []

  [cmbn]
    type = CombinerGenerator
    inputs = 'rectangle_body plane_rigid'
  []

  [cmbn_new]
    type = RenameBlockGenerator
    input = cmbn
    old_block = '0 1'
    new_block = 'box plane'
  []
[]

[Variables]
    [disp_x]
    []
    [disp_y]
    []
    [disp_z]
    []
[]

[Functions]
  [ydisp]
    type = ParsedFunction
    vars = 't0   t_end  disp'
    vals = '10.0  20.0    -4'
    value = 'if (t<t0, 0.0, if (t<t_end, disp*(t-t0)/(t_end-t0), disp))'
  []
  [xdisp]
    type = ParsedFunction
    vars = 't0   t_end  disp'
    vals = '20.0  30.0    2'
    value = 'if (t<t0, 0.0, disp*(t-t0)/(t_end-t0))'
  []
[]

[AuxVariables]
  [./norm_force_x]
  [../]
  [./norm_force_y]
  [../]  
  [./norm_force_z]
  [../]  
  [./tang_force_x]
  [../]
  [./tang_force_y]
  [../]
  [./tang_force_z]
  [../]
[]

[AuxKernels]
  [./normal_force_x]
    type = PenetrationAux
    variable = norm_force_x
    execute_on = timestep_end
    quantity = normal_force_x
    boundary = box_bottom
    paired_boundary = plane_top
  [../]
  [./normal_force_y]
    type = PenetrationAux
    variable = norm_force_y
    execute_on = timestep_end
    quantity = normal_force_y
    boundary = box_bottom
    paired_boundary = plane_top
  [../]
  [./normal_force_z]
    type = PenetrationAux
    variable = norm_force_z
    execute_on = timestep_end
    quantity = normal_force_z
    boundary = box_bottom
    paired_boundary = plane_top
  [../]
  [./tangential_force_x]
    type = PenetrationAux
    variable = tang_force_x
    execute_on = timestep_end
    quantity = tangential_force_x
    boundary = box_bottom
    paired_boundary = plane_top
  [../]
  [./tangential_force_y]
    type = PenetrationAux
    variable = tang_force_y
    execute_on = timestep_end
    quantity = tangential_force_y
    boundary = box_bottom
    paired_boundary = plane_top
  [../]
  [./tangential_force_z]
    type = PenetrationAux
    variable = tang_force_z
    execute_on = timestep_end
    quantity = tangential_force_z
    boundary = box_bottom
    paired_boundary = plane_top
  [../]
[]

[Modules/TensorMechanics/Master]
  [./all]
    add_variables = true
    strain = FINITE
  [../]
[]

[BCs]
  [xfixed]
    type = DirichletBC
    variable = disp_x
    boundary = 'plane_top plane_bottom'
    value = 0
  []
  [yfixed]
    type = DirichletBC
    variable = disp_y
    boundary = 'plane_top plane_bottom'
    value = 0
  []
  [zfixed]
    type = DirichletBC
    variable = disp_z
    boundary = 'plane_top plane_bottom box_top'
    value = 0
  []
  [ydisp]
    type = FunctionDirichletBC
    variable = disp_y
    boundary = box_top
    function = ydisp
    preset = 0
  []
  [xdisp]
    type = FunctionDirichletBC
    variable = disp_x
    boundary = box_top
    function = xdisp
    preset = 0
  []
[]

[Materials]
  [strain_all]
    type = ComputeFiniteStrainElasticStress
    block = 'box plane'
  []

  [Top]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 5
    poissons_ratio = 0.49
    block = 'box'
  []
  [Bottom]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 1e6
    poissons_ratio = 0.3
    block = 'plane'
  []

[]

[Contact]
  [contact_interface]
    primary = plane_top
    secondary = box_bottom
    formulation = penalty
    model = coulomb
    friction_coefficient = 0.1
    normalize_penalty = true
    penalty = 1e+2
    normal_smoothing_distance = 0.1
  []
[]

[Dampers]
  [contact_slip]
    type = ContactSlipDamper
    secondary = box_bottom
    primary = plane_top
    min_damping_factor = 1e-5
  []
[]
[Preconditioning]
  [SMP]
    type = SMP
    full = true
  []
[]

[Executioner]
  type = Transient
  solve_type = NEWTON
  
  petsc_options_iname = '-pc_type -pc_factor_mat_solver_type'
  petsc_options_value = 'lu        mumps'
  line_search = contact
  automatic_scaling = true
  contact_line_search_ltol = 1e-3
  nl_max_its = 100
  l_max_its = 40
  start_time = 0.0
  dt = 0.5
  end_time = 30.0

  verbose = true
  nl_rel_tol = 1e-4
  nl_abs_tol = 1e-4
[] # Executioner

[Outputs]
  color = false
  [out]
    type = Exodus
  []
[] # Outputs
```
![image](https://github.com/idaholab/moose/assets/135039570/eccf880a-2e46-4c51-a551-f32f6cb1fcbf)

![image](https://github.com/idaholab/moose/assets/135039570/72f5b3ba-6d0f-4bfa-aba2-bbdb61193377)

When I look at the output, I see that the tangential forces oscillate unphysically. In the figure below, the average tangential force for the contact surface of the box is shown over time. From time 12.5-20s, the contact is established by pressing the box. After 20s, the box is moved in x-direction. The oscillations in x-component of tangential force is clearly unphysical.

![image](https://github.com/idaholab/moose/assets/135039570/398858f6-6f75-44ff-af1e-79ddab629e56)

However, when I further looked in the code of MechanicalContactConstraint.C in the function ```computeContactForce```, the frictional capacity for the penalty formulation has some different formulation than the literature. 

```
// Frictional capacity
// const Real capacity( _friction_coefficient * (pen_force * pinfo->_normal < 0 ?
// -pen_force * pinfo->_normal : 0) );
const Real capacity(_friction_coefficient *
             (res_vec * pinfo->_normal > 0 ? res_vec * pinfo->_normal : 0));
```

It seems that the current frictional capacity for penalty method is actually the one for kinematic formulation. Although the formulation from literature is also written there, but it is commented. Is there a reasoning why the formulation for kinematic method is used in penalty method?

Also, when I uncomment the actual penalty formulation for friction capacity, the unphysical oscillations in the results disappear. Is this a bug here?

**Comment by lindsayad** [comment]
@bwspenc @recuero for node-face contact

**Comment by recuero** [comment]
It's hard to know exactly what's happening in the lines of code for your problem. Just a few thoughts:

1. You say results are unphysical. Even thought that may be true, what's the expected output for that simulation? It helps to know what to expect for a simple test problem.
2. >the frictional capacity for the penalty formulation has some different formulation than the literature: 

Is the frictional capacity the issue there? I.e. do you get a tangential frictional vector norm larger than the friction coefficient times the normal force?

3. > when I uncomment the actual penalty formulation for friction capacity, the unphysical oscillations in the results disappear. Is this a bug here?

Do you mean you then get the expected results? You can get non-oscillating results with the wrong physics. Also, I'd try increasing the penalty value to try to obtain closer-to-theory results. Maybe @bwspenc can provide some details on the algorithm.

**Reply by npurwar93** [reply]
1. The expected output for this simulation would be: a symmetrical deformation of the box and symmetric tangential force distribution at the contact surface when the box is pressed; when it is horizontally moved, at least the direction of the tangential force at the contact surface should be consistent according to the direction of the motion of the box.

Currently, as visible in the time plot shown above, the direction of the tangential force in x-direction is changing its direction constantly, even when the block is just pressed. Here, I have again plotted (Fig. 1) the x-component of the tangential force at the center node of the contact surface. In Fig. 2, there is no symmetric x-component tangential force distribution at the contact surface. 

Figure 1
![image](https://github.com/idaholab/moose/assets/135039570/fdada590-bf33-4092-9c8c-d34660a6b5b8)

Figure 2
![image](https://github.com/idaholab/moose/assets/135039570/52d99735-7e4f-4e8c-8e61-f3dc17cc46c4)


2. Then magnitude of the tangential force vector is not larger than the friction coefficient times the normal force.  However, the direction is changing every time step.
3.  When I use the following formulation for computing frictional capacity
  ```
const Real capacity( _friction_coefficient * (pen_force * pinfo->_normal < 0 ?
 -pen_force * pinfo->_normal : 0) );
```
I get the expected result. The box first deforms with tangential forces developing symmetrically when the box is pressed (Fig. 3) and then maintains consistent direction when it is moved horizontally (as shown in figure 4 below after 20s).

Figure 3
![image](https://github.com/idaholab/moose/assets/135039570/b0af1bb1-34c7-4892-9217-b52c381340a7)

Figure 4
![image](https://github.com/idaholab/moose/assets/135039570/7f6b47e3-afc6-4dda-9cf3-8244198ca4f2)

Regarding increasing the penalty factor, the original frictional capacity formulation still gives these oscillations, but the time step decreases a lot to achieve convergence. With the new (modified) formulation as in point 3 here, after increasing the penalty factor beyond certain value, the Jacobian becomes negative, probably due to very large deformation causing negative volume elements. It is understandable here that for large penalty factor, the frictional capacity would increase accordingly. And for a soft material as used here, it would cause numerical problems.

**Reply by recuero** [reply]
I was thinking more along the lines of an analytical solution, but I do agree that, qualitatively, the results with the commented code look more physical. It seems to me that for a penalty method you don't need to use the cached residuals (contact forces) to compute a tangential traction, but I am not fully aware of the details of the algorithms used in node-face.

> Regarding increasing the penalty factor, the original frictional capacity formulation still gives these oscillations, but the time step decreases a lot to achieve convergence. With the new (modified) formulation as in point 3 here, after increasing the penalty factor beyond certain value, the Jacobian becomes negative, probably due to very large deformation causing negative volume elements. It is understandable here that for large penalty factor, the frictional capacity would increase accordingly. And for a soft material as used here, it would cause numerical problems.

That makes sense to me. I mentioned increasing the penalty factor because, if selected too small, it can produce very unphysical results.

**Reply by recuero** [reply, marked as answer]
We'll look into this behavior @npurwar93 (https://github.com/idaholab/moose/issues/24632). Thanks for bringing it up.
